cv/lane_detection/src/lane_detection_inferencer.cpp

```diff
@@ -0,0 +1,242 @@
+    for (int r = r_min; r < r_max; ++r) {
+        for (int c = c_min; c < c_max; ++c) {
+            output.emplace_back(input.at<uchar>(r, c));
+        }
+    }
+    return output;
+}
+
+cv::Mat getQuadrantEdges(cv::Mat grayscaled_image, const int row_start,
+                         const int row_end, const int col_start,
+                         const int col_end)
+{
+    const std::vector<uint8_t> quad{convertToQuadrant(
+        grayscaled_image, row_start, row_end, col_start, col_end)};
+
+    const double median{static_cast<double>(getMedian(quad))};
+    const double lower_bound{std::max(0.0, std::floor((1 - 0.205) * median))};
+    const double upper_bound{std::min(255.0, std::floor((1 + 0.205) * median))};
+    cv::Mat quad_cv_mat{
+        cv::Mat(quad, CV_8UC1)
+            .reshape(1, static_cast<int>(grayscaled_image.rows / 2))};
+    cv::Mat edges;
+    cv::Canny(quad_cv_mat, edges, lower_bound, upper_bound);
+    return edges;
+}
+
+cv::Mat getEdgeChannel(cv::Mat grayscaled_image)
+{
+    cv::Mat quad1_edges{getQuadrantEdges(
+        grayscaled_image, 0, static_cast<int>(grayscaled_image.rows / 2), 0,
+        static_cast<int>(grayscaled_image.cols / 2))};
+    cv::Mat quad2_edges{getQuadrantEdges(
+        grayscaled_image, 0, static_cast<int>(grayscaled_image.rows / 2),
+        static_cast<int>(grayscaled_image.cols / 2), grayscaled_image.cols)};
+    cv::Mat quad3_edges{getQuadrantEdges(
+        grayscaled_image, static_cast<int>(grayscaled_image.rows / 2),
+        grayscaled_image.rows, static_cast<int>(grayscaled_image.cols / 2),
+        grayscaled_image.cols)};
+    cv::Mat quad4_edges{getQuadrantEdges(
+        grayscaled_image, static_cast<int>(grayscaled_image.rows / 2),
+        grayscaled_image.rows, 0, static_cast<int>(grayscaled_image.cols / 2))};
+
+    // stitch edges together
+    const int new_rows{std::max(quad1_edges.rows + quad3_edges.rows,
+                                quad2_edges.rows + quad4_edges.rows)};
+    const int new_cols{std::max(quad1_edges.cols + quad3_edges.cols,
+                                quad2_edges.cols + quad4_edges.cols)};
+
+    cv::Mat edges_mask = cv::Mat::zeros(new_rows, new_cols, CV_8UC1);
+
+    quad1_edges.copyTo(
+        edges_mask(cv::Rect(0, 0, quad1_edges.cols, quad1_edges.rows)));
+    quad2_edges.copyTo(edges_mask(
+        cv::Rect(quad1_edges.cols, 0, quad2_edges.cols, quad2_edges.rows)));
+    quad4_edges.copyTo(edges_mask(
+        cv::Rect(0, quad1_edges.rows, quad3_edges.cols, quad3_edges.rows)));
+    quad3_edges.copyTo(
+        edges_mask(cv::Rect(quad1_edges.cols, quad1_edges.rows,
+                            quad4_edges.cols, quad4_edges.rows)));
+    return edges_mask;
+}
+
+cv::Mat getInverseEdgeChannel(cv::Mat edge_channel)
+{
+    cv::Mat inverse_edge_channel;
+    cv::bitwise_not(edge_channel, inverse_edge_channel);
+    return inverse_edge_channel;
+}
+
+cv::Mat combineImages(cv::Mat normalized_img, cv::Mat normalized_edges,
```

Comment:
In hindsight we should be able to combine this function and the `convertHWC2CHW` function

We split the channels, then merge them back together lthen split them again and then concatenate them I think lol

@spencerteetaert I don't rmr why we did this way anymore, was there a reason ?

Reply:
Ease of implementation. The convert HWC2CHW function provided the numpy .permute equivelant for a 3D image and we never looked at merging it with the other concat functions we were doing. Good catch, probably most of these matrix copying/concatenating operations can be reduced.

Reply:
We can optimize later, as you mentioned in a separate comment, performance was good enough